substrate-node/pallets/pallet-smart-contract/src/billing.rs

```diff
@@ -209,418 +343,420 @@ impl<T: Config> Pallet<T> {
+            overdrafted_amount: contract_payment_state.get_overdrafted(),
+        });
+        Ok(().into())
+    }
 
-            // First, distribute extra cultivation rewards if any
-            if contract_lock.has_extra_amount_locked() {
+    // Orchestrate the distribution of rewards
+    // Emits RewardDistributed event
+    // No-Op if contract nither in deleted state nor the distribution frequency is reached
+    fn remit_funds(
+        contract: &types::Contract<T>,
+        contract_payment_state: &mut types::ContractPaymentState<BalanceOf<T>>,
+        src_twin: &pallet_tfgrid::types::Twin<T::AccountId>,
+        farmer_twin: Option<pallet_tfgrid::types::Twin<T::AccountId>>,
+        pricing_policy: &pallet_tfgrid::types::PricingPolicy<T::AccountId>,
+    ) -> DispatchResultWithPostInfo {
+        let is_deleted = matches!(contract.state, types::ContractState::Deleted(_));
+        let should_distribute_rewards =
+            contract_payment_state.cycles >= T::DistributionFrequency::get() || is_deleted;
+        if should_distribute_rewards && contract_payment_state.has_reserved_amount() {
+            let standard_rewards = contract_payment_state.standard_reserved;
+            let additional_rewards = contract_payment_state.additional_reserved;
+            // distribute additional rewards to the farm twin
+
+            let reminder = if let types::ContractData::RentContract(_) = &contract.contract_type {
                 log::info!(
-                    "twin balance {:?} contract lock extra amount {:?}",
-                    twin_balance,
-                    contract_lock.extra_amount_locked
+                    "Distributing additional rewards from twin {:?} with amount {:?}",
+                    src_twin.id,
+                    additional_rewards,
                 );
+                let dst_twin = farmer_twin.ok_or(Error::<T>::TwinNotExists)?;
+                Self::transfer_reserved(
+                    &src_twin.account_id,
+                    &dst_twin.account_id,
+                    additional_rewards,
+                )
+            } else {
+                BalanceOf::<T>::zero()
+            };
 
-                match Self::distribute_extra_cultivation_rewards(
-                    &contract,
-                    twin_balance.min(contract_lock.extra_amount_locked),
-                ) {
-                    Ok(_) => {}
-                    Err(err) => {
-                        log::error!(
-                            "error while distributing extra cultivation rewards {:?}",
-                            err
-                        );
-                        return Err(err);
-                    }
-                };
-
-                // Update twin balance after distribution
-                twin_balance = Self::get_usable_balance(&twin.account_id);
+            let distributed_additional_amount = additional_rewards.saturating_sub(reminder);
+            if reminder > BalanceOf::<T>::zero() {
```

Comment:
Maybe I miss something here but if we have a remainder here it means we were not able to transfer `additional_rewards` because not enough reserve, correct?
So why are we continuing and try to `distribute_standard_rewards()`?

Also why this remainder check is not performed in `fn distribute_standard_rewards()` where `Self::transfer_reserved()` is called ?

Reply:
In theory, what we have in reserve (standard + additional) in `contract_payment_state` should have already been reserved with a call to `<T as Config>::Currency::reserve(...) ` so we should not have remainder , correct?
Have you experienced something different to handle this remainder ?

Reply:
> In theory, what we have in reserve (standard + additional) in contract_payment_state should have already been reserved with a call to <T as Config>::Currency::reserve(...)  so we should not have remainder after a call to <T as Config>::Currency::repatriate_reserved(), correct?

Exactly 

> Have you experienced something different to handle this remainder ?

No, not at all.
The reason I handled the remainder in distribute_additional_rewards is it didn’t require much code. However, I then felt that handling this scenario in the rest of the code might be over-engineering since we aren’t anticipating these cases. So, I made sure that transfer_reserved logs any inconsistencies and left it at that point.

Let me explain in detail.
First, the repatriate_reserved() method in the balances pallet is polite. It uses best efforts during the transfer process and returns any remainder, which is inherent to the transfer_reserved() method.

I was considering whether to ignore the returned remainder, log it, handle it in some way, or fail the transaction in such cases (which is unlikely to happen due to the new reservation logic).

Additionally, I needed to consider whether to handle situations where the destination account is dead. Should this gracefully fail (the dead account not receiving its share), or should it cause the entire transaction to fail? 

I also considered handling the creation of these accounts, allowing transfer_reserved() to perform its work in two steps instead of one. First, we would make use of slash_reserved() to force slash the amount (addressing lock issues, even though we don't currently have logic to add locks), creating an imbalance. Then, we would resolve the imbalance by depositing the same amount into the destination account using resolve_creating (addressing issues with inactive accounts by creating it if needed, although this may not be an expected case).

I was uncertain about the necessity of this additional hardening and I chose not to implement it in the interest of simplicity and efficiency (one operation vs two per beneficiary), but it continued to bother me.

**I made some progressive changes after rethinking this again:**

in commit 4b2d2b0, transfer_reserved() no longer returns the remainder if any; instead, it returns an error if the account is dead or if there is a remainder.
Consequently, remit_fund() will propagate the error, with the only exception being if the contract is in a deleted state. In such cases, it logs a warning and continues the execution anyway.


Adjusted again in commit a1af14a to harden it as previously mentioned by utilizing both slash_reserved() and resolve_creating(), which would eliminate the likelihood of encountering issues related to locked user balances or inactive accounts (if we think this could be happening). This makes the logic bulletproof against any external condition, such as when some logic is introduced to lock some amount of user balance, or when the destination account balance may go below its existence deposit.
On the other side, instead of performing the necessary work in one operation "repatriate_reserved," we perform now two separate operations: "slash" and then "deposit." Each operation results in an event (balances pallet internals).

Both solutions would work.
Please let me know if you think this additional hardening is worthwhile.

Reply:
I understand the dilemma...
We should be bulletproof as much as possible but with less overhead as possible.

> Additionally, I needed to consider whether to handle situations where the destination account is dead. Should this gracefully fail (the dead account not receiving its share), or should it cause the entire transaction to fail?

I think we should not break the entire transaction flow if there is issue with a single transfer.
This is what your first commit proposition commit https://github.com/threefoldtech/tfchain/commit/4b2d2b0e39e740b84616c90110fcd9319a8c2afc does, correct ?

So with your brilliant explanation I ll go with this first commit.
I guess it is enough for now since remainder and dead account should not happen.
We can observe how it behaves and see in practice if we need to add extra hardening in case some un-handled scenarios occurs.

The good thing is that the commit https://github.com/threefoldtech/tfchain/commit/a1af14a1ac19cccdaee546a9a7777d6e10aedec5 is already there for that purpose.

Reply:
Let's consider using a hybrid approach that allows us to benefit from the additional hardening while minimizing overhead.
The repatriate_reserved function is efficient when the beneficiary account exists, so we can use it as the default approach, but fall back to the defensive approach when a dead account is detected.

This would make the transfer almost error-proof with minimal additional overhead (check account existence).
See this commit af76619 for more details.